### ISSUE TITLE
Infinite loop on certain inputs

### ISSUE THREAD

**Issue by aiba** [timothypratley/munkres#5]
`munkres/minimize-weight` seems to hit an infinite loop on the following input.

```clj
(def data
  [[75152.16 79641.735 79816.84000000001
    80136.4100000001 87019.64000000007 93109.72 109210.77200000006
    111735.02900000008 112814.97999999992 117877.14200000005 120774.21500000004
    124174.37499999999 126795.98 133209.9759999998 155223.27800000008
    158271.3189999998 174048.19139999987 174524.84499999986 177995.27399999998
    193350.9690000001 196368.0950000001 228067.74650000012 244482.8307000004
    245169.75620000024 256820.01349999986 298628.8729999993 333048.0404999997]
   [45844.029999999984 65922.10999999993
    71459.125 73218.12999999995 75252.81999999999 76037.58499999999
    77553.08000000005 80057.86999999994 81400.48999999998 82949.50500000002
    86137.78499999996 86765.12499999999 92750.70499999994 93943.25500000002
    102377.82000000005 118454.66500000002 122760.0509999999 131769.749
    132993.62900000016 133541.91 137481.78000000003 143030.455 155900.04799999972
    166807.519 183102.82200000013 185771.57000000007 187268.5659999999
    202959.76099999977 216737.1260000001 217640.96199999985 221026.77900000016]])

(let [[as bs] data]
  (munkres/minimize-weight #(Math/abs ^double (- %1 %2))
                           as bs))
```

Is there some assumption about the input data that I am perhaps violating?  I noticed that in `jstack`, it's stuck at:

```
        at DenseDijkstraSP.<init>(DenseDijkstraSP.java:36)
        at AssignmentProblemDense.augment(AssignmentProblemDense.java:68)
        at AssignmentProblemDense.<init>(AssignmentProblemDense.java:43)
        at munkres$minimize_weight.invokeStatic(munkres.clj:42)
        at munkres$minimize_weight.invoke(munkres.clj:34)
```

And the javadoc for DenseDijkstraSP says "No parallel edges".  Could I be passing in a parallel edge, and if so, how do I avoid that?

**Comment by timothypratley** [comment]
Hi Aiba,


Sorry I don't know the answer to your question.

The java code comes from the "Algorithms" book: https://algs4.cs.princeton.edu/home/
eg: https://algs4.cs.princeton.edu/65reductions/AssignmentProblemDense.java.html

I can see that the code has changed a bit, so it might be worth just getting the latest java code.

If possible, could you download the new java code, and use lein install to test whether it helps?


Regards,
Timothy

**Comment by aiba** [comment]
I suspect it's this "parallel edge" comment in `DenseDijkstraSP.java`, which is still present in the updated code.

Since my weights are fairly large, I tried adding a small deterministic epsilon to the cost function:

```
(defn epsilon ^double [s]
  (/ (double (hash s))
     (double Integer/MAX_VALUE)))

(defn cost [a b]
  (+ (epsilon [a b])
     (Math/abs ^double (- a b))))

(let [[as bs] data]
  (munkres/minimize-weight cost as bs))
```

And indeed, this solves the issue on this input data (and all my other data).

I'm not sure what one would do if one couldn't spare the precision in the cost function.

Tomorrow I'll try it with the latest java code from that Algorithsm book.

**Comment by aiba** [comment]
I'm having trouble tracking down the latest version of AssignmentProblemDense.java.  I did find the official maven repo that tracks the latest algs4 code:

https://github.com/kevin-wayne/algs4

But it doesn't include the "dense" implementation of the assignment problem.  It does have `AssignmentProblem.java`, and when I ran my test case using that, it works without the infinite loop issue.

The documented runtime of `AssignmentProblem.java` is N^3 log(N), vs. just N^3 for `AssignmentProblemDense.java`, but  on this test case, they both ran in ~3ms on my machine, which is plenty fast for my use case.  And given that it solves the infinite loop issue without having to add my own epsilon to the cost function, I prefer it.

I'll submit a PR in case your'e interested in having these changes in this project, otherwise I could depend directly on the algs4 repo and wrap `AssignmentProblem.java` in my own project.

Thanks again for publishing this library.
